File: src/link.ts
// Port of internal/excalidraw/link.go

export type Kind = 'json' | 'room';

export interface Link {
  url: string;
  id: string;
  key: string;
  kind: Kind;
}

export function apiUrl(link: Link): string {
  return link.kind === 'room'
    ? `https://room.excalidraw.com/api/v2/${link.id}`
    : `https://json.excalidraw.com/api/v2/${link.id}`;
}

// Same regex as urlRe in link.go
const URL_RE = /https?:\/\/excalidraw\.com\/#(json|room)=([a-zA-Z0-9]+),([a-zA-Z0-9_\-]+)/g;

/**
 * Returns all unique Excalidraw URLs that have at least one unannotated occurrence.
 * When refresh=true, already-annotated URLs are also returned.
 */
export function parseAll(content: string, refresh: boolean): Link[] {
  const seen = new Set<string>();
  const links: Link[] = [];

  URL_RE.lastIndex = 0;
  let match: RegExpExecArray | null;

  while ((match = URL_RE.exec(content)) !== null) {
    const url = match[0];
    if (seen.has(url)) continue;
    seen.add(url);

    // Idempotency check on the first occurrence (mirrors link.go)
    const afterPos = match.index + url.length;
    let offset = 0;
    if (afterPos < content.length && content[afterPos] === ')') {
      offset = 1; // closing paren of a markdown link
    }
    const trimmed = content.slice(afterPos + offset).replace(/^[ \t]+/, '');
    if (!refresh && trimmed.startsWith('([[')) {
      continue; // already annotated — skip
    }

    links.push({
      url,
      kind: match[1] as Kind,
      id: match[2],
      key: match[3],
    });
  }

  return links;
}

File: cli/internal/excalidraw/link.go
package excalidraw

import (
	"regexp"
	"strings"
)

type Kind string

const (
	JSON Kind = "json"
	Room Kind = "room"
)

// Link is a parsed Excalidraw URL found in a markdown file.
type Link struct {
	URL  string
	ID   string
	Key  string
	Kind Kind
}

// APIURL returns the HTTP endpoint to fetch the scene data from.
func (l Link) APIURL() string {
	switch l.Kind {
	case Room:
		return "https://room.excalidraw.com/api/v2/" + l.ID
	default:
		return "https://json.excalidraw.com/api/v2/" + l.ID
	}
}

var urlRe = regexp.MustCompile(
	`https?://excalidraw\.com/#(json|room)=([a-zA-Z0-9]+),([a-zA-Z0-9_\-]+)`,
)

// ParseAll returns all unique Excalidraw URLs that have at least one
// unannotated occurrence in content.
// When refresh is true, URLs already annotated with ([local copy are also
// included so they can be re-downloaded and their annotation updated.
func ParseAll(content string, refresh bool) []Link {
	seen := map[string]bool{}
	var links []Link

	for _, m := range urlRe.FindAllStringSubmatchIndex(content, -1) {
		url := content[m[0]:m[1]]
		if seen[url] {
			continue
		}
		seen[url] = true

		// Check if this first occurrence is already annotated.
		// injectOne handles each occurrence independently, so this is just
		// used to avoid triggering a download for fully-processed files.
		afterPos := m[1]
		offset := 0
		if afterPos < len(content) && content[afterPos] == ')' {
			offset = 1 // closing paren of a markdown link
		}
		trimmed := strings.TrimLeft(content[afterPos+offset:], " \t")
		if !refresh && strings.HasPrefix(trimmed, "([[") {
			continue // already successfully downloaded
		}

		links = append(links, Link{
			URL:  url,
			Kind: Kind(content[m[2]:m[3]]),
			ID:   content[m[4]:m[5]],
			Key:  content[m[6]:m[7]],
		})
	}

	return links
}
